src/vs/workbench/contrib/workspace/browser/workspace.contribution.ts

```diff
@@ -200,6 +200,32 @@ export class WorkspaceTrustRequestHandler extends Disposable implements IWorkben
 			}
 		}));
 
+		this._register(this.workspaceContextService.onDidChangeWorkspaceFolders(async e => {
+			const trustState = this.workspaceTrustService.getWorkspaceTrustState();
+
+			if (trustState === WorkspaceTrustState.Trusted && e.added.length > 0) {
+				const addedFoldersTrustStateInfo = e.added.map(folder => this.workspaceTrustService.getFolderTrustStateInfo(folder.uri));
+				if (!addedFoldersTrustStateInfo.map(i => i.trustState).every(trustState => trustState === WorkspaceTrustState.Trusted)) {
+					// Adding untrusted/unspecified content to a trusted workspace
+					const result = await this.dialogService.confirm({
+						message: localize('addWorkspaceFolderMessage', "Do you trust the files in this folder?"),
+						detail: localize('addWorkspaceFolderDetail', "You are adding files to a trusted workspace that are not currently trusted. Do you want to trust the new files?\n\nIf you do not trust the new files, we will have to reload the window in order to transition the trust state of the workspace."),
+						primaryButton: localize('yes', 'Yes'),
+						secondaryButton: localize('no', 'No')
+					});
+
+					// Apply changes for the added folders
+					const addedFoldersTrustState = result.confirmed ? WorkspaceTrustState.Trusted : WorkspaceTrustState.Untrusted;
+					addedFoldersTrustStateInfo.map(i => this.workspaceTrustService.setFolderTrustState(i.uri, addedFoldersTrustState));
+
+					return;
+				}
+			}
+
+			// Refresh the workspace trust state
+			this.workspaceTrustService.refreshWorkspaceTrustState();
```

Comment:
why is this necessary?

Reply:
Today, we only recalculate the workspace trust state in case there are changes to the underlying model (list of trusted/untrusted folders). Assume that you have a workspace that is currently `unspecified` and you add a folder that is `untrusted`. While there is no functional difference between the two, we need to "refresh" the workspace trust state so that the editor/status bar correctly shows the trust state.